python/pyspark/serializers.py

```diff
@@ -184,27 +184,67 @@ def loads(self, obj):
         raise NotImplementedError
 
 
-class ArrowSerializer(FramedSerializer):
+class BatchOrderSerializer(Serializer):
     """
-    Serializes bytes as Arrow data with the Arrow file format.
+    Deserialize a stream of batches followed by batch order information.
     """
 
-    def dumps(self, batch):
+    def __init__(self, serializer):
+        self.serializer = serializer
+        self.batch_order = None
+
+    def dump_stream(self, iterator, stream):
+        return self.serializer.dump_stream(iterator, stream)
+
+    def load_stream(self, stream):
+        for batch in self.serializer.load_stream(stream):
+            yield batch
+        num = read_int(stream)
+        self.batch_order = []
+        for i in xrange(num):
+            index = read_int(stream)
+            self.batch_order.append(index)
+        raise StopIteration()
```

Comment:
@BryanCutler, I think this will be broken in Python 3.7 (see [PEP 479](https://www.python.org/dev/peps/pep-0479)). Shall we just remove this line or explicitly `return`?

Reply:
Oh ok, I'm not sure.. I'll look into that.

Reply:
this seems important...